File: skel/R/f_branin.R
##' Branin function.
##'
##' @param x Parameter vector.
##' @return Value of the Branin function.
##' @export
f_branin <- function(x) {
  stopifnot(length(x) == 2)
  x1 <- x[1]
  x2 <- x[2]
  a <- 1
  b <- 5.1 / (2*pi)^2
  c <- 5 / pi
  d <- 6
  e <- 10
  f <- 1 / (8 * pi)
  a * (x2 - b * x1^2 + c * x1 - d)^2 + e * (1 - f) * cos(x1) + e
}

class(f_branin) <- c("branin_function", "soo_function")

##' @S3method lower_bounds branin_function
##' @method lower_bounds branin_function
##' @rdname bounds.Rd
lower_bounds.branin_function <- function(x, dim) {
  stopifnot(dim == 2)
  c(-5, 0)
}

##' @S3method upper_bounds branin_function
##' @method upper_bounds branin_function
##' @rdname bounds.Rd
upper_bounds.branin_function <- function(x, dim) {
  stopifnot(dim == 2)
  c(10, 15)
}

##' @S3method global_minimum branin_function
##' @method global_minimum branin_function
##' @rdname global_minimum.Rd
global_minimum.branin_function <- function(x, dim, ...)
  list(par=list(c(-pi, 12.275), c(pi, 2.275), c(3*pi, 2.475)),
       value=0.397887)

File: src/f_branin.c
#include "soobench.h"

const double f_branin(const double *x, const size_t n) {
  const double a = 1;
  const double b = 5.1 / ((2.0 * M_PI) * (2.0 * M_PI));
  const double c = 5.0 / M_PI;
  const double d = 6.0;
  const double e = 10.0;
  const double f = 1 / (8 * M_PI);
  const double x1 = x[0];
  const double x2 = x[1];
  const double c1 = (x2 - b * x1 * x1 + c * x1 - d);
  return a * c1 * c1 + e * (1 - f) * cos(x1) + e;
}
